### PR TITLE
Utility script for producing inputs to Nerfstudio

### PR DESCRIPTION
**Pull request by womackj1**
After doing a GTSfM reconstruction on the set1_lund_door dataset, you can produce Nerfstudio inputs with the following command:

`python3 prepare_nerfstudio.py --results_path=../../results --camera_model=perspective --images_dir=../../tests/data/set1_lund_door/images/
`

Then, if you have Nerfstudio setup, you can train a nerf with a command like:

`ns-train nerfacto --data nerfstudio-input/ --vis viewer --viewer.websocket-port=7008`

Screen Shot 2022-11-16 at 9 01 29 PM

### CODE REVIEW

File: gtsfm/utils/prepare_nerfstudio.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import argparse
+import os
+import json
+import numpy as np
+from enum import Enum
+from pathlib import Path
+
+
+from gtsfm.utils import images, io
+import thirdparty.nerfstudio.colmap_utils as colmap_utils
```
Comment (ayushbaid) [inline gtsfm/utils/prepare_nerfstudio.py:18]: We follow a specific ordering of imports:

```
<included in python by default>

<third party imports (e.g. numpy)>

<imports from the same package>
```

Can you follow this structure?

File: gtsfm/utils/prepare_nerfstudio.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        output_dir: Path to the output directory.
+        camera_model: Camera model used.
+    Returns:
+        The number of registered images.
+    """
+
+    cameras = colmap_utils.read_cameras_text(cameras_path)
+    images = colmap_utils.read_images_text(images_path)
+
+    # Only supports one camera
```
Comment (ayushbaid) [inline gtsfm/utils/prepare_nerfstudio.py:48]: Does this mean all the cameras should share the same params?

File: gtsfm/utils/prepare_nerfstudio.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    for _, im_data in images.items():
+        rotation = colmap_utils.qvec2rotmat(im_data.qvec)
+        translation = im_data.tvec.reshape(3, 1)
+        w2c = np.concatenate([rotation, translation], 1)
+        w2c = np.concatenate([w2c, np.array([[0, 0, 0, 1]])], 0)
+        c2w = np.linalg.inv(w2c)
+        # Convert from COLMAP's camera coordinate system to ours
+        c2w[0:3, 1:3] *= -1
+        c2w = c2w[np.array([1, 0, 2, 3]), :]
+        c2w[2, :] *= -1
```
Comment (ayushbaid) [inline gtsfm/utils/prepare_nerfstudio.py:61]: I prefer you use GTSAM types and use `.inverse()` on those. Less things to test as we can be more confident in GTSAM's implementation.

File: gtsfm/utils/prepare_nerfstudio.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                "k1": float(camera_params[4]),
+                "k2": float(camera_params[5]),
+                "k3": float(camera_params[6]),
+                "k4": float(camera_params[7]),
+            }
+        )
+
+    out["frames"] = frames
+
+    print(out)
```
Comment (ayushbaid) [inline gtsfm/utils/prepare_nerfstudio.py:102]: Please remove print.

File: gtsfm/utils/prepare_nerfstudio.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        translation = im_data.tvec.reshape(3, 1)
+        w2c = np.concatenate([rotation, translation], 1)
+        w2c = np.concatenate([w2c, np.array([[0, 0, 0, 1]])], 0)
+        c2w = np.linalg.inv(w2c)
+        # Convert from COLMAP's camera coordinate system to ours
+        c2w[0:3, 1:3] *= -1
+        c2w = c2w[np.array([1, 0, 2, 3]), :]
+        c2w[2, :] *= -1
+
+        name = Path(f"./images/{im_data.name}")
```
Comment (ayushbaid) [inline gtsfm/utils/prepare_nerfstudio.py:63]: Which path is this?

File: gtsfm/utils/prepare_nerfstudio.py (hunk 4 above)
Comment (womackj1) [reply gtsfm/utils/prepare_nerfstudio.py:102]: Removed.

File: gtsfm/utils/prepare_nerfstudio.py (hunk 5 above)
Comment (womackj1) [reply gtsfm/utils/prepare_nerfstudio.py:63]: These paths are the paths to the (resized) images that are used as input to nerfstudio.

File: gtsfm/utils/prepare_nerfstudio.py (hunk 2 above)
Comment (womackj1) [reply gtsfm/utils/prepare_nerfstudio.py:48]: Yes, as Nerfstudio's parser does: https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/process_data/colmap_utils.py#L595

File: gtsfm/utils/prepare_nerfstudio.py (hunk 1 above)
Comment (womackj1) [reply gtsfm/utils/prepare_nerfstudio.py:18]: Yes, the imports are now reordered

**Comment by johnwlambert** [comment]
Hi Jonathan, I believe this test is failing:

FAILED tests/utils/test_io_utils.py::TestIoUtils::test_round_trip_cameras_txt - AssertionError: 0.0 != 100.0

**Comment by johnwlambert** [comment]
Closing, since superseded by [https://github.com/borglab/gtsfm/pull/593](https://github.com/borglab/gtsfm/pull/642/files) now.

**Pull request closed without merging**